### PR TITLE
Elaborate array and tuple relation errors

### LINKED ISSUE #30839: Poor error message elaborations for array-like types
```ts
let point = [3, 4] as const;

function distanceFromOrigin([x, y]: [number, number]) {
    return Math.sqrt(x ** 2 + y ** 2);
}

distanceFromOrigin(point);
```

```
Argument of type 'readonly [3, 4]' is not assignable to parameter of type '[number, number]'.
  Type 'readonly [3, 4]' is missing the following properties from type '[number, number]': pop, push, reverse, shift, and 6 more.
```

image

This is bad to show to beginners. Here's a few things we could be doing better.

1. If a source type is a `ReadonlyArray` (for all `E`), and a target type is an `Array`, then give a specialized error message that `ReadonlyArray`s can't be assigned to `Array`s.

   ```
   A 'ReadonlyArray' cannot be assigned to an 'Array' because 'Array's can be mutated.
   ```

2. If the source type is a `readonly` tuple, and the target is a plain tuple or array-like type...

   ```
   A 'readonly' tuple cannot be assigned to a mutable array-like type.
   ```

3. If one side is a tuple type, and the other isn't array-like, just don't elaborate at all. Nobody cares about missing `push`, `pop`, and 6 other members...

### PR DESCRIPTION
**Pull request by jack-williams**
Fixes #30839

### CODE REVIEW

File: tests/baselines/reference/readonlyArraysAndTuples.errors.txt [hunk 1]
```diff
@@ -52,7 +52,7 @@ tests/cases/conformance/types/tuple/readonlyArraysAndTuples.ts(25,5): error TS27
 !!! error TS2740: Type 'readonly string[]' is missing the following properties from type '[string, string]': 0, 1, pop, push, and 5 more.
         mt = rt;  // Error
         ~~
-!!! error TS2740: Type 'readonly [string, string]' is missing the following properties from type '[string, string]': pop, push, reverse, shift, and 3 more.
+!!! error TS4105: A 'readonly' tuple cannot be assigned to a mutable array-like type.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/readonlyArraysAndTuples.errors.txt]: Hmm, I think the fact that these lack context is a bit of a problem. If these aren't acting as elaborations, then maybe both messages could be replaced with

```
The type '{0}' is considered 'readonly' and cannot be assigned to the mutable type '{1}'.
```

Comment (jack-williams) [reply tests/baselines/reference/readonlyArraysAndTuples.errors.txt]: Ah yes, it slipped past me that these errors don't include  `Foo is not assignable  to Bar`. So:
  - `The type '{0}' is considered 'readonly' and cannot be assigned to the mutable type '{1}'.`

which can probably subsume both new error messages?

Comment (RyanCavanaugh) [reply tests/baselines/reference/readonlyArraysAndTuples.errors.txt]: Can we remove the word "considered" here?

Comment (jack-williams) [reply tests/baselines/reference/readonlyArraysAndTuples.errors.txt]: Works for me!

**Comment by DanielRosenwasser** [comment]
Beautiful, thanks Jack!

**Comment by jack-williams** [comment]
Anytime!

**Pull request merged**